### ISSUE TITLE
Why is older fastercsv slower than newer fastercsv

### ISSUE THREAD

**Issue by enebo** [jruby/jruby#7604]
The csv in JRuby 1.7 is quite a bit faster than the one shipped in 9.3/9.4:

```text
system ~/work/jruby master 2052% jruby  -I../snippets ~/work/snippets/csv_bench.rb 
2.4891145080327988
2.0067069232463837
1.8442632555961609
1.718497270718217
1.716303520835936
1.8347881203517318
1.7011602241545916
1.701060843653977
1.7657083729282022
1.8220716370269656
2.243274923413992
```

With current csv:

```text
jruby  -I../snippets ~/work/snippets/csv_bench.rb 
4.355965277180076
3.160095911473036
2.868028214201331
2.769463988021016
3.0816614385694265
3.970997073687613
3.5069410810247064
3.3267764318734407
3.9052677769213915
3.463081202469766
3.283401306718588
3.3949629701673985
```

So almost 2x slower.  MRI 3.1 is maybe 10% slower with the newer versions. Benchmark:

```ruby
require 'csv'
#require 'old_csv'
require 'benchmark'

dir = "data"
Dir.mkdir(dir) unless Dir.exist?(dir)

input = "#{dir}/extract.csv"

unless File.exist?(input)
  Dir.chdir(dir) do
    zip = "sirene_2017111_E_Q.zip"
    system "wget", "-O", zip, "http://files.data.gouv.fr/sirene/#{zip}"
    system "unzip", zip
    system "head -10000 sirc* > extract.csv"
  end
end

30.times do
  p Benchmark.realtime {
    CSV.foreach(input, col_sep: ';', encoding: "ISO-8859-1:UTF-8") { |row|
    }
  }
end
```

**Comment by headius** [comment]
A big difference between 1.7 and every JRuby version that followed is support for string encodings, which under some circumstances can be much more overhead than the old encodingless strings.

I will take a poke at this and see if anything obvious jumps out.

**Comment by headius** [comment]
Indeed, when profiling the top items I see are mostly related to character transcoding.

Modifying your script from

```ruby
CSV.foreach(input, col_sep: ';', encoding: "ISO-8859-1:UTF-8")
```

to

```ruby
CSV.foreach(input, col_sep: ';', encoding: "ISO-8859-1")
```

Nearly doubles the performance on JRuby 9.4.1 and exceeds the performance of JRuby 1.7.

* JRuby 1.7 with "UTF-8": 0.93s
* JRuby 1.7 without "UTF-8": 0.89s
* JRuby 9.4.1 with "UTF-8": 1.30s
* JRuby 9.4.1 without "UTF-8": 0.70s

This doesn't excuse the UTF-8 transcoding being slow, since CRuby does not appear to see such an impact, but it may give you a hint to improve this performance.

**Comment by enebo** [comment]
Just a note on this issue is that @headius was comparing performance of CSV while running JRuby1.7 vs 9.4 but this bug is about why the csv shipped with 1.7 is faster than the csv in 9.4 when running both on 9.4.  Both implementations are receiving the ```encoding: ``` keyword.  The older one I can see pass ```encoding:``` to File.open so it should also be transcoding in the older version of csv.

**Comment by enebo** [comment]
I updated the script above to allow running in either mode via the command line.  You need to download and save the linked csv from 1.7 as old_csv.rb in the same directory as the benchmark script.

**Comment by enebo** [comment]
Yeah both are mysteries which perhaps are related or not:

old_csv on 1.7
```
0.695
0.677
0.682
0.672
```

old_csv on 9.4 on split_opts branch:
```
2.381481336429715
1.7589577538892627
1.5705529348924756
1.5506588239222765
1.6620264453813434
1.552856793627143
```

but then using 9.4 with current csv:
```
2.7935966830700636
2.7348122345283628
2.730583473108709
2.7095523243770003
3.0916133131831884
```

So 1.7 is faster then 9.4 with older csv AND (this issue) the two versions of the same program are slower on 9.4.  The difference in perf on MRI for this issue is a small difference.  Lots o mystery and I suppose both avenues are reasonable here if we want to do that.

**Comment by headius** [comment]
Running on master today, to investigate this issue, I noticed that a longer run of the benchmark eventually optimizes new csv pretty close to the speed of old_csv:

```
[] jruby $ jruby -Xcompile.invokedynamic bench_csv.rb
Using 9.4 csv
3.226551665982697
2.4227594590047374
1.7391455830074847
... about 50 iters later
1.7921142500126734
1.7042538330133539
1.2053204579860903
1.3517762910050806
1.1930326249857899
1.1928771669918206
1.18620429202565
1.1866123329964466
1.1901381249772385
1.1845460839977022
1.1862071250216104
1.185222750005778
1.1858397919859271
```

With threshold=0 it gets close to this, around 1.22s, after just a few iterations.

So it seems the bulk of the initial slowness is due to JIT delays: the main loop in new csv does not get compiled until after 50 parses.

We're still not as fast on new csv as old_csv but we're much closer than it seems.

**Comment by headius** [comment]
I ran three profiles today while investigating this: 9.4 on new csv, 9.4 on old csv, and 1.7 on old csv.

All runs were `-X+C -Xcompile.invokedynamic`.

Both 9.4 profiles were dominated by character transcoding overhead. With the given compiler flags, the new csv was about 1.15s per iteration and old csv was around 0.94s per iteration.

The 1.7 profile showed no transcoding overhead at all. Most of the overhead is in call sites and String manipulation.

I think it's probably safe to say that 9.4 is slower on both versions because it's doing more transcoding work than 1.7. I do not know whether we can eliminate that work or if it is a cost of better encoding support.

**Comment by headius** [comment]
The benchmark as given sets the encoding to "ISO-8859-1:UTF-8" which leads to the transcoding overhead in 9.4 but no similar degradation in 1.7. It is as if 1.7 is not doing that transcoding.

Removing that overhead from 9.4 puts the numbers as follows: new csv: 0.73s, old csv: 0.53s, 1.7 + old: 0.54

The perf difference between new and old on 9.4 remains almost identical, 0.20s per iteration. The additional overhead that makes new csv slower must be unrelated to the transcoding, so remove that to get a cleaner profile.

**Comment by headius** [comment]
Additional profiling without transcoding playing a factor showed me the top profile item in new csv was related to invokedynamic. Based on that I ran on a newer JDK that has better optimization of indy and handles, and the results are striking:

1.7 with old (Java 11): 0.62s
9.4 with old (Java 11): 0.32s
9.4 with new (Java 11): 0.44s

Jumping to Java 21 breaks 1.7, but numbers for 9.4 continue to improve:

9.4 with old (Java 21): 0.28s

This does not solve the mystery of new vs old but they are both faster than on 1.7.
9.4 with new (Java 21): 0.37s

**Comment by headius** [comment]
Profiling on Java 21 with indy overhead out of the picture, and the top profile items for new csv are mostly related to StringScanner and its use of regular expressions:

```
Method	Count	Percentage
org.joni.ByteCodeMachine.executeSb(boolean)	1,034	24.7 %
org.joni.StackMachine.stackInit()	411	9.83 %
org.jruby.ext.strscan.RubyStringScanner.scan(ThreadContext, IRubyObject)	371	8.88 %
org.jruby.util.ByteList.equal(ByteList)	248	5.93 %
java.lang.ThreadLocal$ThreadLocalMap.getEntryAfterMiss(ThreadLocal, int, ThreadLocal$ThreadLocalMap$Entry)	218	5.22 %
org.jruby.RubyInteger.singleCharString(Ruby, byte, RubyClass, Encoding)	190	4.55 %
org.jruby.util.StringSupport.codeRangeScan(Encoding, byte[], int, int)	119	2.85 %
org.jruby.util.collections.WeakValuedMap.get(Object)	110	2.63 %
java.lang.invoke.DirectMethodHandle$Holder.invokeStatic(Object, Object, Object, Object, Object, Object, Object)	103	2.46 %
Users.headius.work.jruby.lib.ruby.stdlib.csv.parser.RUBY$method$parse_quoted_column_value$90(ThreadContext, StaticScope, IRubyObject, Block, RubyModule, String)	103	2.46 %
```

Compare with the old csv profile:

```
Method	Count	Percentage
org.jruby.RubyInteger.singleCharString(Ruby, byte, RubyClass, Encoding)	520	14.4 %
org.jruby.util.ByteList.equal(ByteList)	249	6.92 %
org.jruby.RubyString.substr19(Ruby, int, int)	212	5.89 %
org.jruby.RubyRegexp.preparePattern(RubyString)	187	5.19 %
org.jruby.util.StringSupport.searchNonAscii(byte[], int, int)	183	5.08 %
org.jruby.RubyString.count(ThreadContext, IRubyObject)	137	3.81 %
org.joni.Matcher.searchCommon(int, int, int, int, boolean)	135	3.75 %
org.jruby.RubyString.multiplyByteList(ThreadContext, IRubyObject)	134	3.72 %
java.lang.ThreadLocal$ThreadLocalMap.getEntry(ThreadLocal)	134	3.72 %
org.joni.Matcher.searchInterruptible(int, int, int)	121	3.36 %
```

StrincScanner might not be very efficient with its use of regex, but to not see regex overhead in the old csv profile **at all** is really surprising.

**Comment by enebo** [comment]
Without transcoding we seem to be same speed or better as MRI but it begs the question whether we are doing more transcoding or less efficient transcoding than MRI.  I think the explanation for 1.7 is we just were not transcoding but I guess I cannot say that for 100% sure.  The codebases are so far apart it is probably worth tracing through MRI than 1.7 at this point.

Leaving open for this last piece of yarn to tug on.
